### PR TITLE
Add options to wildcard

### PR DESCRIPTION
**Pull request by CKGrafico**
Extend https://github.com/Realytics/fork-ts-checker-webpack-plugin/pull/77#issuecomment-356566649 PR
In progress adding @Toilal request

### CODE REVIEW

**Comment by Toilal** [comment]
In fact it should read tsconfig `baseUrl` and `paths` properties. In my use case, I have 2 wildcards ... 

```
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@/*": ["src/*"],
      "#/*": ["app/*"]
    },
   ...
  }
}
```

(Library template, source files lies in `src/*` and demo app files lies in `app/*`)

**Comment by CKGrafico** [comment]
Docs modified

**Comment by CKGrafico** [comment]
mm @Toilal this will need more work let me..

**Comment by Toilal** [comment]
If it's not possible, maybe the option could accept an array ?

**Comment by Toilal** [comment]
`ts.CompilerOptions` should contain the `baseUrl` and `paths` properties. It should be possible to support this nicely without any additional configuration. Anyway thanks 👍

**Comment by Toilal** [comment]
Would it be possible to loop on `options.paths` items matching `.*/\*` (regex) until if finds a file for the requested `moduleName` ?

**Comment by CKGrafico** [comment]
I have to change docs and two more minor things and it's done

**Comment by johnnyreilly** [comment]
Good work!  Could you take a look at the failing build please?

**Comment by CKGrafico** [comment]
@johnnyreilly sorry I'm modifiyng my VueProgram.js and translating to your VueProgram.ts xDD

PS: mmm the test always the test haha give me some minutes

**Comment by CKGrafico** [comment]
Your tests are not working for me on Windows because absoluth path returns `C:/...` but I hope are ok now

**Comment by johnnyreilly** [comment]
For what it's worth, I'm a Windows guy but I use the Linux subsystem to run fork-ts-checker-webpack-plugin tests. @piotr-oles could you review this when you get a moment please?

**Comment by CKGrafico** [comment]
But can you do it on another PR xD? Is not related with my changes

**Comment by johnnyreilly** [comment]
Sorry - not sure I understand what you meant?  

Either way - tests are passing now :smile:

File: src/VueProgram.ts [hunk 1]
```diff
@@ -31,24 +31,43 @@ class VueProgram {
...
    * The path is resolved with thie given substitution and includes the CompilerOptions.baseUrl (if given).
    * If no paths given in tsconfig, then the default substitution is '[tsconfig directory]/src'.
    * (This is a fast, simplified inspiration of what's described here: https://github.com/Microsoft/TypeScript/issues/5039)
    */
   public static resolveNonTsModuleName(moduleName: string, containingFile: string, basedir: string, options: ts.CompilerOptions) {
     const baseUrl = options.baseUrl ? options.baseUrl : basedir;
-    const pattern = options.paths ? options.paths['@/*'] : undefined;
-    const substitution = pattern ? options.paths['@/*'][0].replace('*', '') : 'src';
-    const isWildcard = moduleName.substr(0, 2) === '@/';
+    const discartedSymbols = ['.', '..', '/'];
```
Comment (prograhammer) [inline src/VueProgram.ts]: Is "discarted" meant to be "discarded" (misspelling)? :smile:

File: README.md [hunk 2]
```diff
@@ -210,8 +210,7 @@ import Hello from '@/components/hello.vue'
 }
 ```
 
-6. The commonly used `@` path wildcard will work if you set up a `baseUrl` and `paths` (in `compilerOptions`) to include `@/*`. If you don't set this, then 
-the fallback for the `@` wildcard will be `[tsconfig directory]/src` (we hope to make this more flexible on future releases):  
+6. It accepts any wildcard in your TypeScript configuration:
```
Comment (prograhammer) [inline README.md:213]: Since we have this, the `wildcard` option in fork-ts-checker-webpack-plugin` was dropped right?

**Comment by prograhammer** [comment]
@CKGrafico Looks good!

But I see you are going down the route I took too (trying to cover for the common cases). Let's hope there are no more folks giving us future wildcard issues before we beg Microsoft for an API in TypeScript to resolve those paths in tsconfig.json (or we'll finally tackle it ourselves!). :fearful:

File: README.md (hunk 2 above)
Comment (prograhammer) [reply README.md:213]: Yeah, it looks like you dropped that option (good!) since I don't see it here in the final file changes. :stuck_out_tongue:

Comment (CKGrafico) [reply README.md:213]: My fault I was using Github "online" haha and every change was a commit

File: src/VueProgram.ts (hunk 1 above)
Comment (CKGrafico) [reply src/VueProgram.ts]: sorry

**Comment by CKGrafico** [comment]
Anybody needs something else?

**Comment by prograhammer** [comment]
Looks good to me to merge! :rocket:

**Comment by Toilal** [comment]
I experience something I can't understand (again :p)... Even without this change, my project is working properly with `#` as alias (It just runs 0.3.0 release). I can catch TSLint errors on some Vue component imported with `#` symbol. 

You can reproduce this by running, this should raise errors in `src/app/Dummy.vue`

```
git clone https://github.com/Toilal/vue-webpack-template.git -b test-#-import test-import
cd test-import
npm i
npm run build
```

**Comment by Toilal** [comment]
@CKGrafico Could you test with #87 (https://github.com/Toilal/fork-ts-checker-webpack-plugin/tree/vue-non-ts-module-simple) ?

It seems that we can handle wildcard module names the same way we handle relative module name (by building the "full" module name). TypeScript is able to solve paths like `/home/toilal/project/test-import/src/components/#/components/Dummy.vue` making manual building from tsconfig just useless.

**Comment by prograhammer** [comment]
@CKGrafico 

I just ran your PR locally `npm run test` and I'm getting 2 fails. Anybody else getting this?

```
 [INTEGRATION] vue
    ✓ should create a Vue program config if vue=true (298ms)
    ✓ should not create a Vue program config if vue=false (239ms)
    ✓ should create a Vue program if vue=true (269ms)
    ✓ should not create a Vue program if vue=false (239ms)
    ✓ should get syntactic diagnostics from Vue program (254ms)
    1) should not find syntactic errors when checkSyntacticErrors is false
    2) should find syntactic errors when checkSyntacticErrors is true
```

**Comment by CKGrafico** [comment]
I'm going to check asap.
@prograhammer Travis is working well isn't it? I had problems running this on windows (and other PRs of this project) but I will doublecheck

**Comment by Toilal** [comment]
@prograhammer I had this issue too. If you are using yarn, run `yarn upgrade`. (that's why yarn.lock has changed on #87)

**Comment by prograhammer** [comment]
In the integration vue test, the `example-wild.vue` also needs a semantic error (this is good to have in the tests). So that line should be updated to:

`  public x: string = 1;`

Then all passes. Weird that Travis didn't catch it?

**Comment by prograhammer** [comment]
So the total is 3 errors. 2 semantic, and 1 syntactic. 

@CKGrafico if you want to update that line in `example-wild.vue` in the integration test to be a semantic error (just like `example.vue`) in this PR then I think all is good.

**Comment by Toilal** [comment]
I'm pretty sure you don't run same TypeScript or TSLint version than the travis build.

**Comment by prograhammer** [comment]
Ah @Toilal gotcha. Let me check that...

**Comment by prograhammer** [comment]
Or maybe webpack is suddenly silent? I'm on my phone and can't get to my laptop til later tonight to check.

**Comment by CKGrafico** [comment]
@Toilal not sure that your PR :/ will work to every SO
@prograhammer I'm testing this with yarn instead of npm and updating yarn.lock asap, sorry for forget yarn I'm not using it since package-lock.json exist

**Comment by CKGrafico** [comment]
Checked if you don't use windows npm test is working well (The windows errors is not related with this PR we can merge this and made another one)

**Comment by Toilal** [comment]
I'll close my PR, it's simpler but only works magically in my project, with yours we are pretty sure it should work everywhere. Thanks a lot, it'll be included on my webpack template fork https://github.com/Toilal/vue-webpack-template when it's released in this library.

**Comment by CKGrafico** [comment]
Test failing after I've updated yarn? amazing

**Comment by CKGrafico** [comment]
Ok in index.spec  something like this

```
console.log(stats.compilation.errors, stats.compilation.errors.length)
```

returns an array of '1' is not number and length 1 and you're expecting to be 1

but on vue.spec you have the same and you are expecting 2, I'm assuming that is a strange mistake? and correcting it

**Comment by prograhammer** [comment]
Yeah I think there might have been some extra Webpack output there, but with the latest package.json lock it seems to have gone away. (BTW, you guys should update to latest Node 8 (now LTS) and NPM 5 because NPM is as fast as Yarn now (and yarn is just another abstraction to learn on top of NPM anyway).

Anyways, all good to merge! :rocket:

**Comment by Toilal** [comment]
From my experience it's still a bit slower. In fact. when there's a package-lock file, I use npm5, else I use Yarn.

**Comment by CKGrafico** [comment]
I think the PR is ready now :)

**Comment by johnnyreilly** [comment]
Cool - @piotr-oles could you review when you get a moment please?

**Comment by Toilal** [comment]
Could you have a look to https://github.com/Realytics/fork-ts-checker-webpack-plugin/issues/88 ? Not sure how it's related ...

**Comment by CKGrafico** [comment]
Please is you're not to merge this soon, tell us and we will made a fork for our projects :D

**Comment by johnnyreilly** [comment]
@piotr-oles could you take a look when you get a moment? @ckgrafico - until that time would you like me to publish it to our makeshift `@next` channel? https://github.com/TypeStrong/fork-ts-checker-webpack-plugin

**Comment by CKGrafico** [comment]
Not the best but sounds like a solution 💃

**Comment by johnnyreilly** [comment]
Give me an hour...

**Comment by johnnyreilly** [comment]
Take a look: https://github.com/typestrong/fork-ts-checker-webpack-plugin

**Comment by CKGrafico** [comment]
@johnnyreilly mmm I think that you're not using my PR xD https://github.com/CKGrafico/fork-ts-checker-webpack-plugin/commit/f9c01f9dc5dac368df6dbd78097abcd28e16f820 --> https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/commit/d6ce069759a82931e7abe4d44b61d4fc2c95682b#diff-23d36b6fe597b409b93d20e36766ff64R31

**Comment by johnnyreilly** [comment]
Yup - I didn't build from the `patch-1` branch.  Doing that now.

**Comment by johnnyreilly** [comment]
Is this better? https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/commit/4fce3bd4720cb0b10c6ea6b3d75a7a313dcbb91e

**Review by piotr-oles — APPROVED** [review]

**Comment by piotr-oles** [comment]
Thank you for your work and sorry for delay :)

**Pull request merged**